src/rdb.c

```diff
@@ -1214,28 +1214,118 @@ ssize_t rdbSaveSingleModuleAux(rio *rdb, int when, moduleType *mt) {
     return io.bytes;
 }
 
-int functionsSaveRio(rio *rdb) {
-    int ret = C_ERR;
+ssize_t rdbSaveFunctions(rio *rdb) {
     dict *functions = functionsGet();
     dictIterator *iter = dictGetIterator(functions);
     dictEntry *entry = NULL;
+    ssize_t written = 0;
+    ssize_t ret;
     while ((entry = dictNext(iter))) {
-        rdbSaveType(rdb, RDB_OPCODE_FUNCTION);
-        functionInfo *fi = dictGetVal(entry);
-        if (rdbSaveRawString(rdb, (unsigned char *) fi->name, sdslen(fi->name)) == -1) goto done;
-        if (rdbSaveRawString(rdb, (unsigned char *) fi->ei->name, sdslen(fi->ei->name)) == -1) goto done;
+        if ((ret = rdbSaveType(rdb, RDB_OPCODE_FUNCTION)) < 0) goto werr;
+        written += ret;
+        functionInfo* fi = dictGetVal(entry);
+        if ((ret = rdbSaveRawString(rdb, (unsigned char *) fi->name, sdslen(fi->name))) < 0) goto werr;
+        written += ret;
+        if ((ret = rdbSaveRawString(rdb, (unsigned char *) fi->ei->name, sdslen(fi->ei->name))) < 0) goto werr;
+        written += ret;
         if (fi->desc) {
-            if (rdbSaveLen(rdb, 1) == -1) goto done; /* desc exists */
-            if (rdbSaveRawString(rdb, (unsigned char *) fi->desc, sdslen(fi->desc)) == -1) goto done;
+            /* desc exists */
+            if ((ret = rdbSaveLen(rdb, 1)) < 0) return ret;
+            written += ret;
+            if ((ret = rdbSaveRawString(rdb, (unsigned char *) fi->desc, sdslen(fi->desc))) < 0) goto werr;
+            written += ret;
         } else {
-            if (rdbSaveLen(rdb, 0) == -1) goto done; /* desc not exists */
+            /* desc not exists */
+            if ((ret = rdbSaveLen(rdb, 0)) < 0) goto werr;
+            written += ret;
         }
-        if (rdbSaveRawString(rdb, (unsigned char *) fi->code, sdslen(fi->code)) == -1) goto done;
+        if ((ret = rdbSaveRawString(rdb, (unsigned char *) fi->code, sdslen(fi->code))) < 0) goto werr;
+        written += ret;
     }
-    ret = C_OK;
-done:
     dictReleaseIterator(iter);
-    return ret;
+    return written;
+
+werr:
+    dictReleaseIterator(iter);
+    return -1;
+}
+
+ssize_t rdbSaveDb(rio *rdb, int dbid, int rdbflags) {
+    dictIterator *di;
+    dictEntry *de;
+    ssize_t written = 0;
+    ssize_t res;
+    long key_count = 0;
+    long long info_updated_time = 0;
+    size_t processed = 0;
```

Comment:
resetting `processed` here is technically incorrect (should be accumulated by the outer function.
but since we're gonna chuck this code in #9788 soon, i don't mind.

i feel the `info_updated_time` var should still be part of the outer function (might not make a huge difference).

but initializing `key_count` here is certainly a bug.

Reply:
`processed` and `info_updated_time` can be in the inner function, worst case the update will be called 1 extra time per db, which seems fine to me and keeps the function isolated. In the future we might call `sendChildInfo` from other parts of the RDB saving, and if we want to throttle this function I think we should do it internally.

`key_count` is a bug, I missed that. To keep the function isolated, I prefer we send the diff `key_count` on each call and not a running sum, the parent can do the sum. Pushing an appropriate fix, let me know what you think.